File: generated_code/GeminiPro/Java/038_SplayTree.java
/**
 * A complete, self-contained Java class that implements an integer set using a Splay Tree.
 * A splay tree is a self-adjusting binary search tree with the additional property
 * that recently accessed elements are quick to access again.
 */
public class SplayTree {

    /**
     * Inner class representing a node in the Splay Tree.
     */
    private class Node {
        int key;
        Node left, right;

        public Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    public SplayTree() {
        this.root = null;
    }

    /**
     * Performs a right rotation on the subtree rooted at y.
     *
     * @param y The root of the subtree to rotate.
     * @return The new root of the subtree after rotation.
     */
    private Node rightRotate(Node y) {
        Node x = y.left;
        y.left = x.right;
        x.right = y;
        return x;
    }

    /**
     * Performs a left rotation on the subtree rooted at x.
     *
     * @param x The root of the subtree to rotate.
     * @return The new root of the subtree after rotation.
     */
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * The main splay operation. Brings the node with the given key (or the last
     * accessed node if the key is not found) to the root of the tree.
     *
     * @param node The root of the current subtree.
     * @param key  The key to search for and splay.
     * @return The new root of the tree after splaying.
     */
    private Node splay(Node node, int key) {
        // Base cases: node is null or key is present at root
        if (node == null || node.key == key) {
            return node;
        }

        // Key lies in the left subtree
        if (key < node.key) {
            // Key is not in the tree, we are done
            if (node.left == null) {
                return node;
            }

            // Zig-Zig (Left-Left)
            if (key < node.left.key) {
                // Recursively bring the key as root of left-left
                node.left.left = splay(node.left.left, key);
                // First rotation for node
                node = rightRotate(node);
            }
            // Zig-Zag (Left-Right)
            else if (key > node.left.key) {
                // Recursively bring the key as root of left-right
                node.left.right = splay(node.left.right, key);
                // Do first rotation for node.left
                if (node.left.right != null) {
                    node.left = leftRotate(node.left);
                }
            }

            // Do second rotation for node (if required)
            return (node.left == null) ? node : rightRotate(node);

        } else { // Key lies in the right subtree
            // Key is not in tree, we are done
            if (node.right == null) {
                return node;
            }

            // Zig-Zag (Right-Left)
            if (key < node.right.key) {
                // Recursively bring the key as root of right-left
                node.right.left = splay(node.right.left, key);
                // Do first rotation for node.right
                if (node.right.left != null) {
                    node.right = rightRotate(node.right);
                }
            }
            // Zig-Zig (Right-Right)
            else if (key > node.right.key) {
                // Recursively bring the key as root of right-right
                node.right.right = splay(node.right.right, key);
                // First rotation for node
                node = leftRotate(node);
            }

            // Do second rotation for node (if required)
            return (node.right == null) ? node : leftRotate(node);
        }
    }

    /**
     * Inserts a key into the Splay Tree. If the key already exists,
     * the node is splayed to the root.
     *
     * @param key The integer key to insert.
     */
    public void insert(int key) {
        // If the tree is empty, create a new root
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree to bring the closest node to the root
        root = splay(root, key);

        // If key is already present, we are done (it's now the root)
        if (root.key == key) {
            return;
        }

        // Otherwise, create a new node and make it the root
        Node newNode = new Node(key);

        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        } else { // key > root.key
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        root = newNode;
    }

    /**
     * Searches for a key in the Splay Tree. If the key is found, the corresponding
     * node is splayed to the root.
     *
     * @param key The integer key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        root = splay(root, key);

        if (root != null && root.key == key) {
            return key;
        }
        return null;
    }

    /**
     * Deletes a key from the Splay Tree.
     *
     * @param key The integer key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // Splay the tree to bring the node to delete (or its parent) to the root
        root = splay(root, key);

        // If the key is not in the tree, do nothing
        if (root.key != key) {
            return;
        }

        // Now, the node to be deleted is the root
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            // If there's no left child, the right subtree becomes the new tree
            root = rightSubtree;
        } else {
            // Splay the left subtree to bring its maximum element to its root
            // The max element in the left subtree is now the new root of the combined tree
            leftSubtree = splay(leftSubtree, key);

            // Attach the original right subtree to the new root
            leftSubtree.right = rightSubtree;
            root = leftSubtree;
        }
    }
}

File: generated_code/GeminiPro/Python/038_SplayTree.py
import sys

class SplayTree:
    """
    A self-contained Splay Tree class that implements a dictionary-like set for integers.
    
    This implementation includes insert, delete, and search operations. The key
    feature of a splay tree is that after any operation (insert, delete, search),
    the accessed node (or its parent/successor) is moved to the root of the tree
    through a series of rotations. This "splaying" operation makes recently
    accessed elements faster to access again.
    """

    class _Node:
        """A node in the splay tree."""
        def __init__(self, key, parent=None, left=None, right=None):
            self.key = key
            self.parent = parent
            self.left = left
            self.right = right

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _left_rotate(self, x):
        """Performs a left rotation on node x."""
        y = x.right
        x.right = y.left
        if y.left:
            y.left.parent = x
        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation on node x."""
        y = x.left
        x.left = y.right
        if y.right:
            y.right.parent = x
        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
        y.right = x
        x.parent = y

    def _splay(self, x):
        """
        Performs the splay operation on node x, moving it to the root.
        """
        while x.parent:
            p = x.parent
            g = p.parent
            if not g:  # Zig case
                if x == p.left:
                    self._right_rotate(p)
                else:
                    self._left_rotate(p)
            elif x == p.left and p == g.left:  # Zig-Zig case
                self._right_rotate(g)
                self._right_rotate(p)
            elif x == p.right and p == g.right:  # Zig-Zig case
                self._left_rotate(g)
                self._left_rotate(p)
            elif x == p.right and p == g.left:  # Zig-Zag case
                self._left_rotate(p)
                self._right_rotate(g)
            else:  # Zig-Zag case
                self._right_rotate(p)
                self._left_rotate(g)

    def search(self, key):
        """
        Searches for a key in the tree.
        
        If the key is found, the corresponding node is splayed to the root and
        True is returned. If the key is not found, the last accessed node (the
        would-be parent) is splayed to the root and False is returned.

        Args:
            key (int): The integer key to search for.

        Returns:
            bool: True if the key is found, False otherwise.
        """
        if not self.root:
            return False

        current = self.root
        last_visited = None
        while current:
            last_visited = current
            if key < current.key:
                current = current.left
            elif key > current.key:
                current = current.right
            else:
                # Key found, splay the node
                self._splay(current)
                return True
        
        # Key not found, splay the last visited node (parent)
        if last_visited:
            self._splay(last_visited)
        return False

    def insert(self, key):
        """
        Inserts a key into the tree.
        
        If the key is not already in the tree, it is inserted and the new node
        is splayed to the root. If the key already exists, the existing node
        is splayed to the root.

        Args:
            key (int): The integer key to insert.
        """
        # Case 1: Tree is empty
        if not self.root:
            self.root = self._Node(key)
            return

        # Case 2: Tree is not empty, find insertion point
        current = self.root
        parent = None
        while current:
            parent = current
            if key < current.key:
                current = current.left
            elif key > current.key:
                current = current.right
            else:
                # Key already exists, splay it and return
                self._splay(current)
                return
        
        # Key does not exist, insert new node
        new_node = self._Node(key, parent=parent)
        if key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node
        
        # Splay the new node to the root
        self._splay(new_node)

    def delete(self, key):
        """
        Deletes a key from the tree.
        
        If the key is found, it is deleted. The parent of the deleted node (or its
        in-order successor if it had two children) is splayed to the root. If the
        key is not found, the last accessed node during the search is splayed.

        Args:
            key (int): The integer key to delete.
        """
        if not self.root:
            return

        # Splay the node with the key (or its parent if not found) to the root
        self.search(key)
        
        # If the key is not at the root after splaying, it wasn't in the tree
        if self.root.key != key:
            return

        # Now the node to delete is at the root.
        # Split the tree into two: L (left subtree) and R (right subtree)
        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # If no left subtree, the right subtree becomes the new tree
            self.root = right_subtree
            if self.root:
                self.root.parent = None
        elif not right_subtree:
            # If no right subtree, the left subtree becomes the new tree
            self.root = left_subtree
            if self.root:
                self.root.parent = None
        else:
            # Both subtrees exist. Join them.
            # 1. Make the left subtree the current tree.
            self.root = left_subtree
            self.root.parent = None
            
            # 2. Find the maximum element in the left subtree.
            max_node = self.root
            while max_node.right:
                max_node = max_node.right
            
            # 3. Splay this max element. It becomes the new root of the merged tree
            #    and will have no right child.
            self._splay(max_node)
            
            # 4. Attach the original right subtree as the right child of the new root.
            self.root.right = right_subtree
            right_subtree.parent = self.root

    def inorder_traversal(self):
        """Helper method to return keys in-order for verification."""
        result = []
        self._inorder_recursive(self.root, result)
        return result

    def _inorder_recursive(self, node, result):
        if node:
            self._inorder_recursive(node.left, result)
            result.append(node.key)
            self._inorder_recursive(node.right, result)

    def __str__(self):
        """String representation of the tree (in-order)."""
        return str(self.inorder_traversal())

if __name__ == '__main__':
    # Example Usage
    st = SplayTree()
    
    # Insert operations
    print("--- Inserting keys ---")
    keys_to_insert = [10, 20, 30, 40, 50, 25]
    for k in keys_to_insert:
        st.insert(k)
        print(f"Inserted {k}. Root is now: {st.root.key}. Tree: {st}")

    # Search operations
    print("\n--- Searching for keys ---")
    print(f"Search for 25: {st.search(25)}")
    print(f"Root after searching 25: {st.root.key}. Tree: {st}")
    
    print(f"Search for 10: {st.search(10)}")
    print(f"Root after searching 10: {st.root.key}. Tree: {st}")
    
    print(f"Search for 99 (not present): {st.search(99)}")
    print(f"Root after searching 99: {st.root.key}. Tree: {st}") # Parent (50) is splayed

    # Delete operations
    print("\n--- Deleting keys ---")
    print(f"Deleting 30. Current tree: {st}")
    st.delete(30)
    print(f"Root after deleting 30: {st.root.key}. Tree: {st}") # 25 (successor) becomes root
    
    print(f"Deleting 50. Current tree: {st}")
    st.delete(50)
    print(f"Root after deleting 50: {st.root.key}. Tree: {st}") # 40 becomes root
    
    print(f"Deleting 99 (not present). Current tree: {st}")
    st.delete(99)
    print(f"Root after trying to delete 99: {st.root.key}. Tree: {st}") # No change